### PR TITLE
FEATURE: [Experimental] Content Security Policy

### PR DESCRIPTION
**Pull request by xrav3nz**
This introduces four Site Settings:

- `content_security_policy`: Turn on Content-Security-Policy
- `content_security_policy_report_only`: Turn on Content-Security-Policy-Report-Only
- `content_security_policy_collect_reports`: Enable CSP violation report collection at `/csp_reports` (I added this so reporting can be temporarily disabled to avoid flooding the logs)
- `content_security_policy_script_src`: Additional whitelisted script sources. The current host and CDN are included by default.

Right now, this only adds the `script_src` CSP directive, which specifies the valid sources for JavaScripts.

### CODE REVIEW

File: app/controllers/application_controller.rb [hunk 1]
```diff
@@ -501,6 +513,10 @@ def secure_session
 
   private
 
+  def content_security_policy
+    @content_security_policy ||= ContentSecurityPolicy.new.build
```
Comment (xrav3nz) [inline app/controllers/application_controller.rb]: Do we need to cache this so we don't have to calculate it for every request? Note the cache needs to be invalidated when the related site settings update

File: app/controllers/application_controller.rb [hunk 2]
```diff
@@ -101,6 +103,16 @@ def dont_cache_page
...
 
+  def set_content_security_policy_header
+    if SiteSetting.content_security_policy
+      response.headers['Content-Security-Policy'] = content_security_policy
+    end
+
+    if SiteSetting.content_security_policy_report_only
+      response.headers['Content-Security-Policy-Report-Only'] = content_security_policy
+    end
+  end
```
Comment (xrav3nz) [inline app/controllers/application_controller.rb]: Should we skip this when the request is not a `GET HTML`?

File: config/site_settings.yml [hunk 3]
```diff
@@ -1664,40 +1673,40 @@ uncategorized:
...
   default_email_mailing_list_mode_frequency:
-    enum: 'MailingListModeSiteSetting'
+    enum: "MailingListModeSiteSetting"
     default: 1
   disable_mailing_list_mode:
     default: false
     client: true
   default_email_always: false
   default_email_previous_replies:
-    enum: 'PreviousRepliesSiteSetting'
```
Comment (SamSaffron) [inline config/site_settings.yml]: yeah ... lets not do all of these now :)

**Comment by SamSaffron** [comment]
As a security feature I wonder if we are setting it early enough, in particular omniauth is not going to get it now. 

Maybe just swap out: `use ActionDispatch::ContentSecurityPolicy::Middleware` and place our own implementation there ?

**Comment by xrav3nz** [comment]
Argh how did I not see this 😂 will revert

File: lib/content_security_policy.rb [hunk 4]
```diff
@@ -0,0 +1,79 @@
+# frozen_string_literal: true
+require_dependency 'global_path'
+
+class ContentSecurityPolicy
+  include GlobalPath
+
+  class Middleware
+    WHITELISTED_PATHS = %w(
+      /logs
```
Comment (xrav3nz) [inline lib/content_security_policy.rb]: Logster still has [quite some inline JS](https://github.com/discourse/logster/blob/0b4b42fe815b333f56bf7daa804b1aa2fb0642cc/lib/logster/middleware/viewer.rb#L204-L248).

Comment (SamSaffron) [reply lib/content_security_policy.rb]: Ouch ... can you work with @OsamaSayegh on getting that removed asap?

File: lib/content_security_policy.rb [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+    end
+
+    def call(env)
+      request = Rack::Request.new(env)
+      _, headers, _ = response = @app.call(env)
+
+      return response unless html_response?(headers)
+      return response if whitelisted?(request.path)
+
+      policy = ContentSecurityPolicy.new.build
```
Comment (SamSaffron) [inline lib/content_security_policy.rb]: why do any building if this is not enabled?

**Comment by SamSaffron** [comment]
overall this looks fine to me sans my tiny new comment, once you address that first thing in your day can you merge this in and enable this on meta? that way you will have all of friday to watch it. and if you feel any kind of risk you can just turn it off on meta at the end of your day.

**Comment by xrav3nz** [comment]
Sadly had to revert :( more info and new PR: https://github.com/discourse/discourse/pull/6514

**Pull request merged**